api/tokens.go

```diff
@@ -0,0 +1,179 @@
+			return
+		}
+
+		defer req.Body.Close()
+
+		authParams := make(map[string]string)
+		err = json.Unmarshal(body, &authParams)
+		if err != nil {
+			errorMessage := "api endpoint POST login returned an error unmarshalling the body"
+			handleUnexpectedError(ctx, w, err, errorMessage, field, param)
+			return
+		}
+
+		validPasswordRequest := passwordValidation(authParams)
+		validEmailRequest := emailValidation(authParams)
+		if err != nil {
+			errorMessage := "api endpoint POST login returned an error validating the email"
+			handleUnexpectedError(ctx, w, err, errorMessage, field, param)
+			return
+		}
+
+		invalidPasswordError := errors.New("Invalid password")
+		invalidPasswordMessage := "Unable to validate the password in the request"
+		invalidEmailError := errors.New("Invalid email")
+		invalidErrorMessage := "Unable to validate the email in the request"
+
+		invalidPasswordErrorBody := individualErrorBuilder(invalidPasswordError, invalidPasswordMessage, field, param)
+		invalidEmailErrorBody := individualErrorBuilder(invalidEmailError, invalidErrorMessage, field, param)
+
+		var errorList []IndividualError
+
+		if !(validPasswordRequest) && !(validEmailRequest) {
+
+			errorList = nil
+			errorList = append(errorList, invalidPasswordErrorBody)
+			errorList = append(errorList, invalidEmailErrorBody)
+
+			errorResponseBody := errorResponseBodyBuilder(errorList)
+			writeErrorResponse(ctx, w, 400, errorResponseBody)
+			return
+		}
+
+		if !validPasswordRequest {
+
+			errorList = nil
+			errorList = append(errorList, invalidPasswordErrorBody)
+			errorResponseBody := errorResponseBodyBuilder(errorList)
+
+			writeErrorResponse(ctx, w, 400, errorResponseBody)
+			return
+		}
+
+		if !validEmailRequest {
+
+			errorList = nil
+			errorList = append(errorList, invalidEmailErrorBody)
+			errorResponseBody := errorResponseBodyBuilder(errorList)
+
+			writeErrorResponse(ctx, w, 400, errorResponseBody)
+			return
+		}
+	}
+}
+
+func passwordValidation(requestBody map[string]string) (isPasswordValid bool) {
+
+	isPasswordValid = false
+
+	if len(requestBody["password"]) != 0 {
+		isPasswordValid = true
+	}
+
+	return isPasswordValid
+}
+
+//emailValidation checks for both a valid email address and an empty email address
+func emailValidation(requestBody map[string]string) (isEmailValid bool) {
+
+	isEmailValid = false
+	isEmailValid = emailvalidation.IsEmailValid(requestBody["email"])
+
+	return isEmailValid
+}
+
+func individualErrorBuilder(err error, message, sourceField, sourceParam string) (individualError IndividualError) {
+
+	individualError = IndividualError{
+		SpecificError: error.Error(err),
+		Message:       message,
+		Source: Source{
+			Field: sourceField,
+			Param: sourceParam},
+	}
+
+	return individualError
+}
+
+func errorResponseBodyBuilder(listOfErrors []IndividualError) (errorResponseBody ErrorStructure) {
+
+	errorResponseBody = ErrorStructure{
+		Errors: listOfErrors,
+	}
+
+	return errorResponseBody
+}
```

Comment:
Consider moving the Error types and these functions into a separate package. It feels like it is its own separate thing, this will also make it easier if/when we extract this into it's own library,

Reply:
These have been moved to a separate package now